calico_containers/adapter/datastore.py

```diff
@@ -590,6 +596,87 @@ def remove_ip_pool(self, version, pool):
             raise KeyError("%s is not a configured IP pool." % pool)
 
     @handle_errors
+    def get_hostnames(self):
+        """
+        Return a list of hostnames
+        :return:  A list of hostnames configured in etcd.
+        """
+        # Get a list of the hosts.
+        hosts = []
+        results = self.etcd_client.read(HOSTS_PATH)
+        for result in results.children:
+            packed = result.key.split("/")
+            # Handle no hosts being configured (etcd returns the parent).
+            if len(packed) != 5:
+                continue
+            hosts.append(packed[-1])
+        return hosts
+
+    @handle_errors
+    def get_node_as_peer(self, version, hostname):
+        """
+        Return the node BGP information as a BGPPeer object
+
+        :param version: "v4" for IPv4, "v6" for IPv6
+        :param hostname: The node hostname.
+        :return: A BGP Peer object, or None if there is no IP configuration of
+        the requested version.
+        """
+        assert version in ("v4", "v6")
+        if version == "v4":
+            bgp_ip = BGP_NODE_IPV4 %  {"hostname": hostname}
+        else:
+            bgp_ip = BGP_NODE_IPV6 %  {"hostname": hostname}
+        bgp_as = BGP_NODE_AS %  {"hostname": hostname}
+
+        # Get the BGP IP address.  If no key is present, return None.
+        try:
+            ip = self.etcd_client.read(bgp_ip).value
+        except EtcdKeyNotFound:
+            return None
+        if not ip:
+            return None
+
+        # Get the BGP AS number.  If not present, use the default.
+        try:
+            as_num = self.etcd_client.read(bgp_as).value
+        except EtcdKeyNotFound:
+            as_num = None
+        if not as_num:
+            as_num = self.get_default_node_as()
+
+        return BGPPeer(ip, as_num)
+
+    @handle_errors
+    def get_node_to_node_mesh_peers(self, version, hostname):
+        """
+        Get the node-to-node mesh peers for this host.
+
+        :param version: "v4" for IPv4, "v6" for IPv6
+        :param hostname: The node hostname.
+        :return: List of BGPPeer.
+        """
+        assert version in ("v4", "v6")
+        if not self.get_bgp_node_mesh():
+            return []
+
+        our_peer = self.get_node_as_peer(version, hostname)
+        if not our_peer:
+            # We have no BGP information for the requested version and so
+            # we will not peer with other nodes.
+            return []
+
+        peers = []
+        hostnames = self.get_hostnames()
+        for ohostname in hostnames:
+            if ohostname == hostname:
+                continue
+            peer = self.get_node_as_peer(version, ohostname)
```

Comment:
Implementing it this way doubles the number of etcd requests we make, since the `get_hostnames()` already walks the tree once, throwing away all the returned AS & IP data.

Do we have other reasons to want to `get_hostnames()`?